File: src/osc.py
from pythonosc import udp_client, dispatcher, osc_server
from tinyoscquery.queryservice import OSCQueryService
from tinyoscquery.utility import get_open_tcp_port, get_open_udp_port, check_if_tcp_port_open, check_if_udp_port_open
from tinyoscquery.query import OSCQueryBrowser, OSCQueryClient
import time
import os
from threading import Thread
from psutil import process_iter
import logging
from ovr import FINGERS, SPLAYFINGERS
import copy

AVATAR_PARAMETERS_PREFIX = "/avatar/parameters/"
AVATAR_CHANGE_PARAMETER = "/avatar/change"

class OSC:
    def __init__(self, conf: dict, avatar_change_function, run_server = True) -> None:
        self.config = conf
        self.ip = conf["IP"]
        self.port = conf["Port"]
        self.server_port = conf["Server_Port"]
        self.http_port = conf["HTTP_Port"]
        self.stick_tolerance = int(conf['StickMoveTolerance']) / 100
        self.curr_avatar = ""
        self.curr_time = 0
        self.binary_num_bits = int(conf["Binary_bits"])
        self.binary_potencies = [2**i for i in range(self.binary_num_bits)]
        self.binary_potency = (2**self.binary_num_bits) - 1
        self.server = None
        self.oscqs = None
        self.osc_client = udp_client.SimpleUDPClient(self.ip, self.port)
        if run_server:
            self.start_server(avatar_change_function)
        
    def start_server(self, avatar_change_function) -> None:
        """
        Starts the OSC server and OSCQuery endpoints and advertises them.
        Parameters:
            avatar_change_function (function): Function to be called when the avatar changes
        Returns:
            None
        """
        self.disp = dispatcher.Dispatcher()
        self.disp.map(AVATAR_CHANGE_PARAMETER, avatar_change_function)
        if self.server_port != 9001:
            logging.info("OSC Server port is not default, testing port availability and advertising OSCQuery endpoints")
            if self.server_port <= 0 or not check_if_udp_port_open(self.server_port):
                self.server_port = get_open_udp_port()
            if self.http_port <= 0 or not check_if_tcp_port_open(self.http_port):
                self.http_port = self.server_port if check_if_tcp_port_open(self.server_port) else get_open_tcp_port()
        else:
            logging.info("OSC Server port is default.")

        logging.info("Waiting for VRChat to start.")
        while not self.is_running():
            time.sleep(3)
        logging.info("VRChat started!")
        self.qclient = self._wait_get_oscquery_client()
        self.curr_avatar = self.qclient.query_node(AVATAR_CHANGE_PARAMETER).value[0]
        self.server = osc_server.ThreadingOSCUDPServer((self.ip, self.server_port), self.disp)
        server_thread = Thread(target=self._osc_server_serve, daemon=True)
        server_thread.start()
        self.oscqs = OSCQueryService("ThumbParamsOSC", self.http_port, self.server_port)
        self.oscqs.advertise_endpoint(AVATAR_CHANGE_PARAMETER, access="readwrite")


    def is_running(self) -> bool:
        """
        Checks if VRChat is running.
        Returns:
            bool: True if VRChat is running, False if not
        """
        _proc_name = "VRChat.exe" if os.name == 'nt' else "VRChat"
        return _proc_name in (p.name() for p in process_iter())


    def _wait_get_oscquery_client(self) -> OSCQueryClient:
        """
        Waits for VRChat to be discovered and ready and returns the OSCQueryClient.
        Returns:
            OSCQueryClient: OSCQueryClient for VRChat
        """
        service_info = None
        logging.info("Waiting for VRChat to be discovered.")
        while service_info is None:
            browser = OSCQueryBrowser()
            time.sleep(2) # Wait for discovery
            service_info = browser.find_service_by_name("VRChat")
        logging.info("VRChat discovered!")
        client = OSCQueryClient(service_info)
        logging.info("Waiting for VRChat to be ready.")
        while client.query_node(AVATAR_CHANGE_PARAMETER) is None:
            time.sleep(2)
        logging.info("VRChat ready!")
        return client


    def _osc_server_serve(self) -> None:
        """
        Starts the OSC server.
        Returns:
            None
        """
        logging.info(f"Starting OSC client on {self.ip}:{self.server_port}:{self.http_port}")
        self.server.serve_forever(2)

    
    def _float_to_binary(self, float_value):
        """
        Converts a float value to binary parameters.
        Parameters:
            float_value (float): Float value to convert
            num_bits (int): Number of bits to use
        Returns:
            list: Binary parameters + negative flag
        """
        negative = float_value < 0
        float_value = abs(float_value)

        # Calculate the total base-10 value based on the float value and number of bits
        total_base_10 = int(float_value * self.binary_potency)
        
        # Convert the total base-10 value to binary parameters
        binary_params = [int(bit) for bit in format(total_base_10, f'0{self.binary_num_bits}b')]
        
        # Add the negative flag to the end of the binary parameters and return
        return binary_params + [negative]


    def _binary_to_float(self, binary_params: list):
        """
        Reconstructs binary parameters back to a float value.
        Parameters:
            binary_params (list): Binary parameters + negative flag
        Returns:
            float: Float value
        """
        negative = binary_params.pop()
        # Calculate the total base-10 value from the binary parameters
        total_base_10 = int(''.join(map(str, binary_params)), 2)
        
        # Map the total base-10 value to a float in the range [0.0, 1.0]
        float_value = total_base_10 / self.binary_potency

        return -float_value if negative else float_value


    def _send_boolean_toggle(self, action: dict, value: bool) -> None:
        """
        Sends a boolean action as a toggle to VRChat via OSC.
        Parameters:
            action (dict): Action
            value (bool): Value of the parameter
        Returns:
            None
        """
        if not action["enabled"]:
            return
        
        if value:
            action["last_value"] = not action["last_value"]
            time.sleep(0.1)
            self.send_parameter(action["osc_parameter"], action["last_value"])
            return
        
        if action["always"]:
            self.send_parameter(action["osc_parameter"], action["last_value"])


    def _send_boolean(self, action: dict, value: bool) -> None:
        """
        Sends a boolean action to VRChat via OSC.
        Parameters:
            action (dict): Action
            value (bool): Value of the parameter
        Returns:
            None
        """
        _do_send = action["always"] == 2 or (action["always"] == 0 and action["last_value"] != value) or (action["always"] == 1 and value)

        if not action["enabled"] or not _do_send:
            return

        if action["floating"]:
            if value:
                action["timestamp"] = self.curr_time
            elif not value and self.curr_time - action["timestamp"] <= action["floating"]: 
                value = action["last_value"]

        if _do_send:
            self.send_parameter(action["osc_parameter"], value)
            action["last_value"] = value


    def _send_vector1(self, action: dict, value: float) -> None:
        """
        Sends a vector1 action to VRChat via OSC.
        Parameters:
            action (dict): Action
            value (float): Value of the parameter
        Returns:
            None
        """
        _do_send = action["always"] == 2 or (action["always"] == 0 and action["last_value"] != value) or (action["always"] == 1 and value)

        if not action["enabled"] or not _do_send:
            return
        
        if action["floating"]:
            if value > action["last_value"]:
                action["timestamp"] = self.curr_time
            elif value < action["last_value"] and self.curr_time - action["timestamp"] <= action["floating"]: 
                value = action["last_value"]

        if _do_send:
            if action["binary"]:
                tmp = self._float_to_binary(value)
                self.send_parameter(action["osc_parameter"] + "_Negative", tmp[-1])
                tmp = tmp[:-1]
                tmp.reverse()
                for i in range(len(tmp)):
                    self.send_parameter(action["osc_parameter"] + str(self.binary_potencies), tmp[i])
            else:
                self.send_parameter(action["osc_parameter"], value)
            action["last_value"] = value


    def _send_vector2(self, action: dict, value: tuple) -> None:
        """
        Sends a vector2 action to VRChat via OSC.
        Parameters:
            action (dict): Action
            value (tuple): X and Y value of the parameter in a tuple
        Returns:
            None
        """
        val_x, val_y = value[0], value[1]
        if action["unsigned"][0]:
            val_x = (val_x + 1) / 2
        if action["unsigned"][1]:
            val_y = (val_y + 1) / 2

        if action["floating"]:
            if val_x:
                action["timestamp"][0] = self.curr_time
            elif not val_x and self.curr_time - action["timestamp"][0] <= action["floating"][0]: 
                val_x = action["last_value"][0]
            if val_y:
                action["timestamp"][1] = self.curr_time
            elif not val_y and self.curr_time - action["timestamp"][1] <= action["floating"][1]:
                val_y = action["last_value"][1]

        val_bool = (val_x > self.stick_tolerance or val_y > self.stick_tolerance) or (val_x < -self.stick_tolerance or val_y < -self.stick_tolerance)
        value = [val_x, val_y, val_bool]

        for i in range(len(action["osc_parameter"])):
            if not action["enabled"][i]:
                continue
            if action["always"][i] == 2 or (action["always"][i] == 0 and action["last_value"][i] != value[i]) or (action["always"][i] == 1 and value[i]):
                if action["binary"][i]:
                    value[i] = self._float_to_binary(value[i])
                    tmp = value[i]
                    self.send_parameter(action["osc_parameter"][i] + "_Negative", tmp[-1])
                    tmp = tmp[:-1]
                    tmp.reverse()
                    for j in range(len(tmp)):
                        self.send_parameter(action["osc_parameter"][i] + str(self.binary_potencies), tmp[j])
                else:
                    self.send_parameter(action["osc_parameter"][i], value[i])
                action["last_value"][i] = value[i]


    def _send_skeleton(self, action: dict, skeleton) -> None:
        if skeleton is None or not action["enabled"]:
            return

        base_osc_parameter = action['osc_parameter']
        parameters = [(FINGERS, "Curl", skeleton.flFingerCurl), (SPLAYFINGERS, "Splay", skeleton.flFingerSplay)]

        for param_group, param_type, finger_data in parameters:
            for fname, fidx in param_group.items():
                osc_parameter = f"{base_osc_parameter}/{param_type}/{fname}"
                self._send_vector1({**action, "osc_parameter": osc_parameter}, finger_data[fidx])


    def refresh_time(self) -> None:
        """
        Refreshes the current time.
        Returns:
            None
        """
        self.curr_time = time.time()


    def send_parameter(self, parameter: str, value) -> None:
        """
        Sends a parameter to VRChat via OSC.
        Parameters:
            parameter (str): Name of the parameter
            value (any): Value of the parameter
            floating (str): Floating value of the parameter, just for for debug output
        Returns:
            None
        """
        self.osc_client.send_message(AVATAR_PARAMETERS_PREFIX + parameter, value)


    def send(self, action: dict | str, value: bool | float | tuple) -> None:
        """
        Sends an OSC message to VRChat.
        Parameters:
            action (dict | str): Action or OSC address
            value (bool | float | tuple): Value of the parameter
        Returns:
            None
        """
        if isinstance(action, str):
            self.config[action]["osc_parameter"] = action
            self.config[action]["floating"] = 0.0
            self._send_boolean(self.config[action], value)
            return
        
        match action['type']:
            case "boolean":
                if action["floating"] == -1:
                    self._send_boolean_toggle(action, value)
                else:
                    self._send_boolean(action, value)
            case "vector1":
                self._send_vector1(action, value)
            case "vector2":
                self._send_vector2(action, value)
            case "skeleton":
                self._send_skeleton(action, value)
            case _:
                raise TypeError("Unknown action type: " + action['type'])

    def shutdown(self) -> None:
        """
        Stops the OSC server.
        Returns:
            None
        """
        if self.server:
            self.server.shutdown()
        if self.oscqs:
            self.oscqs.stop()

File: src/ThumbParamsOSC/OSC.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace ThumbParamsOSC;

/// <summary>
/// Handles OSC communication with VRChat (client + server) and implements all parameter-send logic.
/// Mirrors the Python OSC class.
/// </summary>
internal sealed class OSC : IDisposable
{
    // ── constants ──────────────────────────────────────────────────────────────
    private const string AVATAR_PARAMETERS_PREFIX = "/avatar/parameters/";
    private const string AVATAR_CHANGE_PARAMETER = "/avatar/change";

    // ── fields ─────────────────────────────────────────────────────────────────
    private readonly JObject _config;
    private UdpClient? _udpClient;
    private UdpClient? _udpServer;
    private readonly Action<string, object>? _avatarChangeCallback;

    private int _binaryNumBits;
    private int[] _binaryPotencies = Array.Empty<int>();
    private int _binaryPotency;
    private float _stickTolerance;

    public readonly string IP;
    public readonly int Port;
    public readonly int ServerPort;
    public readonly int HttpPort;
    public string CurrAvatar = "";
    public double CurrTime;

    // ── constructor ────────────────────────────────────────────────────────────
    public OSC(JObject config, Action<string, object> avatarChangeCallback, bool runServer)
    {
        _config = config;
        _avatarChangeCallback = avatarChangeCallback;

        IP = config["IP"]!.Value<string>()!;
        Port = config["Port"]!.Value<int>();
        ServerPort = config["Server_Port"]!.Value<int>();
        HttpPort = config["HTTP_Port"]!.Value<int>();
        _stickTolerance = (float)(config["StickMoveTolerance"]!.Value<int>() / 100.0);
        _binaryNumBits = config["Binary_bits"]!.Value<int>();
        _binaryPotencies = Enumerable.Range(0, _binaryNumBits).Select(i => (int)Math.Pow(2, i)).ToArray();
        _binaryPotency = (int)Math.Pow(2, _binaryNumBits) - 1;

        _udpClient = new UdpClient();
        _udpClient.Connect(IP, Port);

        if (runServer)
        {
            _ = Task.Run(() => StartServerAsync(avatarChangeCallback));
        }
    }

    // ── time ───────────────────────────────────────────────────────────────────
    public void RefreshTime() =>
        CurrTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

    // ── public send API ────────────────────────────────────────────────────────

    /// <summary>
    /// Sends a named parameter (ControllerType, LeftThumb, etc.) by looking up the config dict.
    /// Mirrors the Python: if isinstance(action, str): … _send_boolean(…).
    /// </summary>
    public void Send(string actionName, object? value)
    {
        if (!_config.ContainsKey(actionName)) return;
        var action = (JObject)_config[actionName]!;
        action["osc_parameter"] = actionName;
        action["floating"] = 0.0;
        SendBoolean(action, value);
    }

    /// <summary>Sends a value for a config action dict, dispatching by type.</summary>
    public void Send(JObject action, object? value)
    {
        string type = action["type"]!.Value<string>()!;
        switch (type)
        {
            case "boolean":
                float floatingVal = action["floating"] is JToken ft ? ft.Value<float>() : 0f;
                if (floatingVal == -1f)
                    SendBooleanToggle(action, value is bool b && b);
                else
                    SendBoolean(action, value);
                break;
            case "vector1":
                SendVector1(action, value is float fv ? fv : Convert.ToSingle(value));
                break;
            case "vector2":
                if (value is (float vx, float vy))
                    SendVector2(action, vx, vy);
                else if (value is ValueTuple<float, float> t2)
                    SendVector2(action, t2.Item1, t2.Item2);
                break;
            case "skeleton":
                SendSkeleton(action, value as SkeletonData);
                break;
        }
    }

    // ── internal send methods ──────────────────────────────────────────────────

    private void SendBoolean(JObject action, object? value)
    {
        bool enabled = action["enabled"]!.Value<bool>();
        int always = action["always"]!.Value<int>();
        var lastValueToken = action["last_value"];
        object lastValue = lastValueToken?.Type == JTokenType.Boolean
            ? (object)lastValueToken.Value<bool>()
            : (lastValueToken?.Value<double>() ?? 0);

        bool doSend = always == 2
            || (always == 0 && !AreEqual(lastValue, value))
            || (always == 1 && IsTruthy(value));

        if (!enabled || !doSend) return;

        float floating = action["floating"] is JToken ft ? ft.Value<float>() : 0f;
        double timestamp = action["timestamp"] is JToken ts ? ts.Value<double>() : 0;

        if (floating > 0f)
        {
            if (IsTruthy(value))
            {
                action["timestamp"] = CurrTime;
            }
            else if (!IsTruthy(value) && CurrTime - timestamp <= floating)
            {
                value = lastValue;
            }
        }

        SendParameter(action["osc_parameter"]!.Value<string>()!, value);
        action["last_value"] = value == null ? null : JToken.FromObject(value);
    }

    private void SendBooleanToggle(JObject action, bool pressed)
    {
        if (!action["enabled"]!.Value<bool>()) return;

        if (pressed)
        {
            bool last = action["last_value"]?.Value<bool>() ?? false;
            bool newVal = !last;
            action["last_value"] = newVal;
            string paramName = action["osc_parameter"]!.Value<string>()!;
            // Delay asynchronously so the polling loop is not blocked.
            _ = Task.Run(async () =>
            {
                await Task.Delay(100);
                SendParameter(paramName, newVal);
            });
            return;
        }

        int always = action["always"]!.Value<int>();
        if (always != 0)
            SendParameter(action["osc_parameter"]!.Value<string>()!, action["last_value"]?.Value<bool>() ?? false);
    }

    private void SendVector1(JObject action, float value)
    {
        bool enabled = action["enabled"]!.Value<bool>();
        int always = action["always"]!.Value<int>();
        float lastValue = action["last_value"] is JToken lv ? lv.Value<float>() : 0f;

        bool doSend = always == 2
            || (always == 0 && !lastValue.Equals(value))
            || (always == 1 && value != 0f);

        if (!enabled || !doSend) return;

        float floating = action["floating"] is JToken ft ? ft.Value<float>() : 0f;
        double timestamp = action["timestamp"] is JToken ts ? ts.Value<double>() : 0;

        if (floating > 0f)
        {
            if (value > lastValue)
                action["timestamp"] = CurrTime;
            else if (value < lastValue && CurrTime - timestamp <= floating)
                value = lastValue;
        }

        bool binary = action["binary"] is JToken bt && bt.Value<bool>();
        if (binary)
        {
            var (bits, negative) = FloatToBinary(value);
            SendParameter(action["osc_parameter"]!.Value<string>()! + "_Negative", negative);
            for (int i = 0; i < bits.Length; i++)
                SendParameter(action["osc_parameter"]!.Value<string>()! + _binaryPotencies[i].ToString(), bits[i] == 1);
        }
        else
        {
            SendParameter(action["osc_parameter"]!.Value<string>()!, value);
        }

        action["last_value"] = value;
    }

    private void SendVector2(JObject action, float valX, float valY)
    {
        var oscParams = (JArray)action["osc_parameter"]!;
        var alwaysArr = (JArray)action["always"]!;
        var enabledArr = (JArray)action["enabled"]!;
        var lastValueArr = (JArray)action["last_value"]!;
        var floatingArr = (JArray)action["floating"]!;
        var timestampArr = (JArray)action["timestamp"]!;
        var unsignedArr = (JArray)action["unsigned"]!;
        var binaryArr = (JArray)action["binary"]!;

        if (unsignedArr[0].Value<bool>()) valX = (valX + 1f) / 2f;
        if (unsignedArr[1].Value<bool>()) valY = (valY + 1f) / 2f;

        float flt0 = floatingArr[0].Value<float>();
        float flt1 = floatingArr[1].Value<float>();
        double ts0 = timestampArr[0].Value<double>();
        double ts1 = timestampArr[1].Value<double>();
        float lv0 = lastValueArr[0].Value<float>();
        float lv1 = lastValueArr[1].Value<float>();

        if (flt0 > 0f)
        {
            if (valX != 0f) timestampArr[0] = CurrTime;
            else if (CurrTime - ts0 <= flt0) valX = lv0;
        }
        if (flt1 > 0f)
        {
            if (valY != 0f) timestampArr[1] = CurrTime;
            else if (CurrTime - ts1 <= flt1) valY = lv1;
        }

        bool valBool = valX > _stickTolerance || valY > _stickTolerance
                    || valX < -_stickTolerance || valY < -_stickTolerance;

        object[] values = { valX, valY, valBool };

        for (int i = 0; i < oscParams.Count; i++)
        {
            if (!enabledArr[i].Value<bool>()) continue;

            int always = alwaysArr[i].Value<int>();
            object lastVal = i < 2
                ? (object)lastValueArr[i].Value<float>()
                : (object)lastValueArr[i].Value<bool>();

            bool doSend = always == 2
                || (always == 0 && !AreEqual(lastVal, values[i]))
                || (always == 1 && IsTruthy(values[i]));

            if (!doSend) continue;

            bool binary = binaryArr[i].Value<bool>();
            if (binary && i < 2)
            {
                float fval = Convert.ToSingle(values[i]);
                var (bits, negative) = FloatToBinary(fval);
                string paramName = oscParams[i].Value<string>()!;
                SendParameter(paramName + "_Negative", negative);
                for (int j = 0; j < bits.Length; j++)
                    SendParameter(paramName + _binaryPotencies[j].ToString(), bits[j] == 1);
                lastValueArr[i] = JToken.FromObject(values[i]);
            }
            else
            {
                SendParameter(oscParams[i].Value<string>()!, values[i]);
                lastValueArr[i] = JToken.FromObject(values[i]);
            }
        }
    }

    private void SendSkeleton(JObject action, SkeletonData? skeleton)
    {
        if (skeleton == null || !action["enabled"]!.Value<bool>()) return;

        string baseParam = action["osc_parameter"]!.Value<string>()!;

        // Finger curls
        var curlNames = new[] { ("Thumb", Fingers.Thumb), ("Index", Fingers.Index), ("Middle", Fingers.Middle), ("Ring", Fingers.Ring), ("Pinky", Fingers.Pinky) };
        foreach (var (name, idx) in curlNames)
            SendVector1(BuildSkeletonSubAction(action, $"{baseParam}/Curl/{name}"), skeleton.FingerCurl[idx]);

        // Finger splays
        var splayNames = new[] { ("Index", SplayFingers.Index), ("Middle", SplayFingers.Middle), ("Ring", SplayFingers.Ring), ("Pinky", SplayFingers.Pinky) };
        foreach (var (name, idx) in splayNames)
            SendVector1(BuildSkeletonSubAction(action, $"{baseParam}/Splay/{name}"), skeleton.FingerSplay[idx]);
    }

    /// <summary>
    /// Creates a copy of the skeleton action with a specific osc_parameter.
    /// Mirrors the Python {**action, "osc_parameter": osc_parameter} dict spread.
    /// </summary>
    private static JObject BuildSkeletonSubAction(JObject action, string oscParameter)
    {
        var clone = (JObject)action.DeepClone();
        clone["osc_parameter"] = oscParameter;
        return clone;
    }

    // ── binary helpers ─────────────────────────────────────────────────────────

    private (int[] bits, bool negative) FloatToBinary(float value)
    {
        bool negative = value < 0;
        value = Math.Abs(value);
        int totalBase10 = (int)(value * _binaryPotency);
        string bitStr = Convert.ToString(totalBase10, 2).PadLeft(_binaryNumBits, '0');
        int[] bits = bitStr.Select(c => c == '1' ? 1 : 0).Reverse().ToArray(); // LSB first
        return (bits, negative);
    }

    // ── raw OSC send ───────────────────────────────────────────────────────────

    /// <summary>
    /// Sends an OSC message to the configured VRChat endpoint.
    /// </summary>
    public void SendParameter(string parameter, object? value)
    {
        string address = AVATAR_PARAMETERS_PREFIX + parameter;
        byte[] packet = BuildOscPacket(address, value);
        try { _udpClient?.Send(packet, packet.Length); }
        catch (Exception ex) { AppLogger.Error($"OSC send error: {ex.Message}"); }
    }

    // ── OSC packet encoding ────────────────────────────────────────────────────

    internal static byte[] BuildOscPacket(string address, object? value)
    {
        using var ms = new MemoryStream();
        WriteOscString(ms, address);

        if (value is bool b)
        {
            WriteOscString(ms, b ? ",T" : ",F");
        }
        else if (value is int intVal)
        {
            WriteOscString(ms, ",i");
            ms.Write(ToBigEndian(intVal));
        }
        else if (value is float floatVal)
        {
            WriteOscString(ms, ",f");
            ms.Write(ToBigEndian(floatVal));
        }
        else if (value is double doubleVal)
        {
            WriteOscString(ms, ",f");
            ms.Write(ToBigEndian((float)doubleVal));
        }
        else if (value is string str)
        {
            WriteOscString(ms, ",s");
            WriteOscString(ms, str);
        }
        else
        {
            // Fallback: try int
            WriteOscString(ms, ",i");
            ms.Write(ToBigEndian(Convert.ToInt32(value)));
        }

        return ms.ToArray();
    }

    private static void WriteOscString(MemoryStream ms, string s)
    {
        byte[] bytes = Encoding.ASCII.GetBytes(s + "\0");
        ms.Write(bytes, 0, bytes.Length);
        int pad = (4 - (bytes.Length % 4)) % 4;
        if (pad > 0) ms.Write(new byte[pad], 0, pad);
    }

    private static byte[] ToBigEndian(int v)
    {
        byte[] b = BitConverter.GetBytes(v);
        if (BitConverter.IsLittleEndian) Array.Reverse(b);
        return b;
    }

    private static byte[] ToBigEndian(float v)
    {
        byte[] b = BitConverter.GetBytes(v);
        if (BitConverter.IsLittleEndian) Array.Reverse(b);
        return b;
    }

    // ── OSC receive / avatar change ────────────────────────────────────────────

    private async Task StartServerAsync(Action<string, object> avatarChangeCallback)
    {
        try
        {
            int udpPort = ServerPort;
            int httpPortNum = HttpPort;

            if (udpPort != 9001)
            {
                AppLogger.Info("OSC Server port is not default, testing port availability.");
                udpPort = udpPort <= 0 || !IsUdpPortOpen(udpPort) ? GetOpenUdpPort() : udpPort;
                httpPortNum = httpPortNum <= 0 || !IsTcpPortOpen(httpPortNum)
                    ? (IsTcpPortOpen(udpPort) ? udpPort : GetOpenTcpPort())
                    : httpPortNum;
            }

            AppLogger.Info("Waiting for VRChat to start.");
            while (!IsVrChatRunning())
                await Task.Delay(3000);
            AppLogger.Info("VRChat started!");

            var qClient = await WaitGetOscQueryClientAsync();
            if (qClient != null)
            {
                string? avatarId = await qClient.QueryAvatarIdAsync();
                if (avatarId != null)
                    CurrAvatar = avatarId;
            }

            _udpServer = new UdpClient(udpPort);
            AppLogger.Info($"OSC server listening on port {udpPort}.");
            _ = Task.Run(() => OscReceiveLoopAsync(avatarChangeCallback));

            var oscQuery = new OscQueryService("ThumbParamsOSC", httpPortNum, udpPort);
            oscQuery.Start();
            AppLogger.Info($"OSCQuery advertised on HTTP port {httpPortNum}.");
        }
        catch (Exception ex)
        {
            AppLogger.Error($"OSC server error: {ex.Message}");
        }
    }

    private async Task OscReceiveLoopAsync(Action<string, object> cb)
    {
        while (_udpServer != null)
        {
            try
            {
                var result = await _udpServer.ReceiveAsync();
                var (address, args) = ParseOscPacket(result.Buffer);
                if (address == AVATAR_CHANGE_PARAMETER && args.Length > 0)
                    cb(address, args[0]);
            }
            catch (ObjectDisposedException) { break; }
            catch { }
        }
    }

    // ── OSC packet decoding ────────────────────────────────────────────────────

    internal static (string address, object[] args) ParseOscPacket(byte[] data)
    {
        int pos = 0;
        string address = ReadOscString(data, ref pos);
        if (pos >= data.Length) return (address, Array.Empty<object>());

        string typeTags = ReadOscString(data, ref pos);
        var args = new List<object>();

        for (int i = 1; i < typeTags.Length; i++)
        {
            switch (typeTags[i])
            {
                case 'i':
                    args.Add(ReadInt32(data, ref pos));
                    break;
                case 'f':
                    args.Add(ReadFloat32(data, ref pos));
                    break;
                case 's':
                    args.Add(ReadOscString(data, ref pos));
                    break;
                case 'T':
                    args.Add(true);
                    break;
                case 'F':
                    args.Add(false);
                    break;
            }
        }

        return (address, args.ToArray());
    }

    private static string ReadOscString(byte[] data, ref int pos)
    {
        int start = pos;
        while (pos < data.Length && data[pos] != 0) pos++;
        string s = Encoding.ASCII.GetString(data, start, pos - start);
        pos++; // null terminator
        int pad = (4 - (pos % 4)) % 4;
        pos += pad;
        return s;
    }

    private static int ReadInt32(byte[] data, ref int pos)
    {
        byte[] b = data[pos..(pos + 4)];
        if (BitConverter.IsLittleEndian) Array.Reverse(b);
        pos += 4;
        return BitConverter.ToInt32(b);
    }

    private static float ReadFloat32(byte[] data, ref int pos)
    {
        byte[] b = data[pos..(pos + 4)];
        if (BitConverter.IsLittleEndian) Array.Reverse(b);
        pos += 4;
        return BitConverter.ToSingle(b);
    }

    // ── port utilities ─────────────────────────────────────────────────────────

    private static int GetOpenUdpPort()
    {
        using var s = new UdpClient(0);
        return ((IPEndPoint)s.Client.LocalEndPoint!).Port;
    }

    private static int GetOpenTcpPort()
    {
        using var s = new System.Net.Sockets.TcpListener(IPAddress.Loopback, 0);
        s.Start();
        int port = ((IPEndPoint)s.LocalEndpoint).Port;
        s.Stop();
        return port;
    }

    private static bool IsUdpPortOpen(int port)
    {
        try { using var s = new UdpClient(port); return true; }
        catch { return false; }
    }

    private static bool IsTcpPortOpen(int port)
    {
        try
        {
            using var s = new System.Net.Sockets.TcpListener(IPAddress.Loopback, port);
            s.Start(); s.Stop(); return true;
        }
        catch { return false; }
    }

    private static bool IsVrChatRunning()
    {
        foreach (var p in System.Diagnostics.Process.GetProcesses())
        {
            if (p.ProcessName.Equals("VRChat", StringComparison.OrdinalIgnoreCase))
                return true;
        }
        return false;
    }

    private static async Task<OscQueryClient?> WaitGetOscQueryClientAsync()
    {
        AppLogger.Info("Waiting for VRChat to be discovered via OSCQuery.");
        OscQueryClient? client = null;
        while (client == null)
        {
            client = await OscQueryBrowser.FindVrChatAsync();
            if (client == null)
                await Task.Delay(2000);
        }
        AppLogger.Info("VRChat discovered!");

        AppLogger.Info("Waiting for VRChat to be ready.");
        while (await client.QueryAvatarIdAsync() == null)
            await Task.Delay(2000);
        AppLogger.Info("VRChat ready!");
        return client;
    }

    // ── helpers ────────────────────────────────────────────────────────────────

    private static bool IsTruthy(object? v) =>
        v switch
        {
            bool b => b,
            int i => i != 0,
            float f => f != 0f,
            double d => d != 0.0,
            null => false,
            _ => true,
        };

    private static bool AreEqual(object? a, object? b)
    {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        if (a is float fa && b is float fb) return Math.Abs(fa - fb) < 1e-6f;
        if (a is double da && b is double db) return Math.Abs(da - db) < 1e-9;
        return a.Equals(b);
    }

    public void Shutdown()
    {
        try { _udpClient?.Close(); } catch { }
        try { _udpServer?.Close(); } catch { }
        _udpClient = null;
        _udpServer = null;
    }

    public void Dispose() => Shutdown();
}
